### PR TITLE
Feature gate CLI flag and visibility changes

### PR DESCRIPTION
**Pull request by Aneurysm9**
Adds CLI flags for setting feature gates, exposes the feature gate registry type and provides an accessor for the global feature gate registry.

### CODE REVIEW

File: service/featuregate/flags.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+var gatesList = FlagValue{}
+
+// Flags adds CLI flags for managing feature gates to the provided FlagSet
+func Flags(flags *flag.FlagSet) {
+	flags.Var(
+		gatesList,
+		gatesListCfg,
+		"Comma-delimited list of feature gate identifiers. Prefix with '-' to disable the feature.  '+' or no prefix will enable the feature.")
+}
```
Comment (bogdandrutu) [inline service/featuregate/flags.go:33]: I would like to see at least couple of examples for the rules. At least in the func comment.

File: service/featuregate/gates.go [hunk 2]
```diff
@@ -27,38 +27,45 @@ type Gate struct {
...
 func Apply(cfg map[string]bool) {
-	reg.apply(cfg)
+	reg.Apply(cfg)
 }
 
-type registry struct {
+// GlobalRegistry returns the Registry instance used by package-global functions in this package.
+func GlobalRegistry() *Registry {
+	return reg
+}
```
Comment (bogdandrutu) [inline service/featuregate/gates.go:59]: Do you need this to be public?

File: service/featuregate/gates.go [hunk 3]
```diff
@@ -27,38 +27,45 @@ type Gate struct {
...
+
+type Registry struct {
```
Comment (bogdandrutu) [inline service/featuregate/gates.go:61]: What is the need in this PR that adds the flag to make this public?

File: service/featuregate/gates.go (hunk 2 above)
Comment (Aneurysm9) [reply service/featuregate/gates.go:59]: [You asked for this to be added.](https://github.com/open-telemetry/opentelemetry-collector/pull/4108#discussion_r726260681)

File: service/featuregate/gates.go (hunk 3 above)
Comment (bogdandrutu) [reply service/featuregate/gates.go:61]: You still did not answer why is this needed.

File: service/featuregate/gates.go (hunk 2 above)
Comment (bogdandrutu) [reply service/featuregate/gates.go:59]: Sure, per the comment there "and have only a "GetRegistry" func that gets the global one". The idea was not to have both global funcs and global Registry, but have only one global func "GetGlobalRegistry" and the Registry itself.

Comment (bogdandrutu) [reply service/featuregate/gates.go:59]: Also that was called a "prototype" and you sneak it in a PR that we need (adding the flag)... not sure this is the right call.

File: service/featuregate/gates.go (hunk 3 above)
Comment (bogdandrutu) [reply service/featuregate/gates.go:61]: I understand where this comes from, but my comment was to remove the other global funcs.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
@Aneurysm9 for reasons like https://github.com/open-telemetry/opentelemetry-collector/pull/4430 we MUST never accept different changes in same PR.

File: service/featuregate/flags.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+var _ flag.Value = (*FlagValue)(nil)
+
+// FlagValue implements the flag.Value interface and provides a mechanism for applying feature
+// gate statuses to a Registry
+type FlagValue map[string]bool
+
+// String returns a string representing the FlagValue
+func (f FlagValue) String() string {
```
Comment (bogdandrutu) [inline service/featuregate/flags.go:49]: Does this need to be public? (I want to understand)

Comment (Aneurysm9) [reply service/featuregate/flags.go:49]: Yes.  It is part of the [`flag.Value` interface](https://pkg.go.dev/flag#Value).

Comment (bogdandrutu) [reply service/featuregate/flags.go:49]: Sorry wrong line of comment, was looking at SetSlice (you can answer on Slack or after the merge).

**Pull request merged**